File: fiberplane-charts/src/Mondrian/generator/utils/calculateYAxisRange.ts
import type { Axis } from "../../types";
import type { Buckets, MinMax } from "../types";
import { getYAxisForConstantValue } from "./getYAxisForConstantValue";

/**
 * Detects the range to display along the Y axis by looking at all the min-max
 * values inside the buckets.
 *
 * When rendering a stacked chart, use `calculateStackedYAxisRange()` instead.
 */
export function calculateYAxisRange<T>(
  buckets: Buckets<T>,
  getMinMax: (bucket: T) => MinMax,
): Axis {
  const minMax = getBucketsMinMax(buckets, getMinMax);
  if (!minMax) {
    return getYAxisForConstantValue(0);
  }

  let [minValue, maxValue] = minMax;

  if (minValue === maxValue) {
    return getYAxisForConstantValue(minValue);
  }

  const distance = maxValue - minValue;
  const margin = 0.05 * distance;

  if (minValue < 0 || minValue >= margin) {
    minValue -= margin;
  } else {
    minValue = 0;
  }

  if (maxValue > 0 || maxValue <= -margin) {
    maxValue += margin;
  } else {
    maxValue = 0;
  }

  return { minValue, maxValue };
}

function getBucketsMinMax<T>(
  buckets: Buckets<T>,
  getMinMax: (bucketValue: T) => MinMax,
): MinMax | undefined {
  let minMax: MinMax | undefined;

  for (const value of buckets.values()) {
    const bucketMinMax = getMinMax(value);
    if (!minMax) {
      minMax = bucketMinMax;
      continue;
    }

    if (bucketMinMax[0] < minMax[0]) {
      minMax[0] = bucketMinMax[0];
    }
    if (bucketMinMax[1] > minMax[1]) {
      minMax[1] = bucketMinMax[1];
    }
  }

  return minMax;
}

File: mondrian-charts/src/fiberplane/utils/calculate_y_axis_range.rs
use super::get_y_axis_for_constant_value;
use crate::fiberplane::{Buckets, MinMax};
use crate::types::Axis;

/// Detects the range to display along the Y axis by looking at all the min-max
/// values inside the buckets.
///
/// When rendering a stacked chart, use
/// [`super::calculate_stacked_y_axis_range()`] instead.
pub(crate) fn calculate_y_axis_range<T: Clone>(
    buckets: &Buckets<T>,
    get_min_max: impl Fn(T) -> MinMax,
) -> Axis {
    let Some(min_max) = get_buckets_min_max(buckets, get_min_max) else {
        return get_y_axis_for_constant_value(0.);
    };

    let MinMax(mut min_value, mut max_value) = min_max;

    if min_value == max_value {
        return get_y_axis_for_constant_value(min_value);
    }

    let distance = max_value - min_value;
    let margin = 0.05 * distance;

    if min_value < 0. || min_value >= margin {
        min_value -= margin;
    } else {
        min_value = 0.;
    }

    if max_value > 0. || max_value <= -margin {
        max_value += margin;
    } else {
        max_value = 0.;
    }

    Axis {
        min_value,
        max_value,
        tick_suggestions: None,
    }
}

fn get_buckets_min_max<T: Clone>(
    buckets: &Buckets<T>,
    get_min_max: impl Fn(T) -> MinMax,
) -> Option<MinMax> {
    let mut min_max: Option<MinMax> = None;

    for value in buckets.values() {
        let bucket_min_max = get_min_max(value.clone());
        min_max = Some(if let Some(min_max) = min_max {
            min_max.extend_with_other(bucket_min_max)
        } else {
            bucket_min_max
        });
    }

    min_max
}
